File: examples/Training/python/ROIs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (C) 2014 University of Dundee & Open Microscopy Environment.
#                    All Rights Reserved.
# Use is subject to license terms supplied in LICENSE.txt
#

"""
FOR TRAINING PURPOSES ONLY!
"""

import numpy
import struct
import math

import omero
from omero.model.enums import UnitsLength
from omero.rtypes import rdouble, rint, rstring
from omero.gateway import BlitzGateway
from omero.gateway import ColorHolder
from Parse_OMERO_Properties import USERNAME, PASSWORD, HOST, PORT
from Parse_OMERO_Properties import imageId
imageId = int(imageId)

"""
start-code
"""

# Create a connection
# ===================
conn = BlitzGateway(USERNAME, PASSWORD, host=HOST, port=PORT)
conn.connect()
updateService = conn.getUpdateService()


# Create ROI
# ==========
# We are using the core Python API and omero.model objects here, since ROIs
# are not yet supported in the Python Blitz Gateway.
#
# First we load our image and pick some parameters for shapes
x = 50
y = 200
width = 100
height = 50
image = conn.getObject("Image", imageId)
z = image.getSizeZ() / 2
t = 0


# We have a helper function for creating an ROI and linking it to new shapes
def create_roi(img, shapes):
    # create an ROI, link it to Image
    roi = omero.model.RoiI()
    # use the omero.model.ImageI that underlies the 'image' wrapper
    roi.setImage(img._obj)
    for shape in shapes:
        roi.addShape(shape)
    # Save the ROI (saves any linked shapes too)
    return updateService.saveAndReturnObject(roi)


# Another helper for generating the color integers for shapes
def rgba_to_int(red, green, blue, alpha=255):
    """ Return the color as an Integer in RGBA encoding """
    r = red << 24
    g = green << 16
    b = blue << 8
    a = alpha
    rgba_int = r+g+b+a
    if (rgba_int > (2**31-1)):       # convert to signed 32-bit int
        rgba_int = rgba_int - 2**32
    return rgba_int


# create a Rectangle shape (added to ROI below)
print ("Adding a rectangle at theZ: %s, theT: %s, X: %s, Y: %s, width: %s,"
       " height: %s" % (z, t, x, y, width, height))
rect = omero.model.RectangleI()
rect.x = rdouble(x)
rect.y = rdouble(y)
rect.width = rdouble(width)
rect.height = rdouble(height)
rect.theZ = rint(z)
rect.theT = rint(t)
rect.textValue = rstring("test-Rectangle")
rect.fillColor = rint(rgba_to_int(255, 255, 255, 255))
rect.strokeColor = rint(rgba_to_int(255, 255, 0, 255))

# create an Ellipse shape (added to ROI below)
ellipse = omero.model.EllipseI()
ellipse.x = rdouble(y)
ellipse.y = rdouble(x)
ellipse.radiusX = rdouble(width)
ellipse.radiusY = rdouble(height)
ellipse.theZ = rint(z)
ellipse.theT = rint(t)
ellipse.textValue = rstring("test-Ellipse")

# Create an ROI containing 2 shapes on same plane
# NB: OMERO.insight client doesn't support display
# of multiple shapes on a single plane.
# Therefore the ellipse is removed later (see below)
create_roi(image, [rect, ellipse])

# create an ROI with single line shape
line = omero.model.LineI()
line.x1 = rdouble(x)
line.x2 = rdouble(x+width)
line.y1 = rdouble(y)
line.y2 = rdouble(y+height)
line.theZ = rint(z)
line.theT = rint(t)
line.textValue = rstring("test-Line")
create_roi(image, [line])


def create_mask(mask_bytes, bytes_per_pixel=1):
    if bytes_per_pixel == 2:
        divider = 16.0
        format_string = "H"  # Unsigned short
        byte_factor = 0.5
    elif bytes_per_pixel == 1:
        divider = 8.0
        format_string = "B"  # Unsigned char
        byte_factor = 1
    else:
        message = "Format %s not supported"
        raise ValueError(message)
    steps = math.ceil(len(mask_bytes) / divider)
    mask = []
    for i in range(long(steps)):
        binary = mask_bytes[
            i * int(divider):i * int(divider) + int(divider)]
        format = str(int(byte_factor * len(binary))) + format_string
        binary = struct.unpack(format, binary)
        s = ""
        for bit in binary:
            s += str(bit)
        mask.append(int(s, 2))
    return bytearray(mask)

mask_x = 50
mask_y = 50
mask_h = 100
mask_w = 100
# Create [0, 1] mask
mask_array = numpy.fromfunction(
    lambda x, y: (x * y) % 2, (mask_w, mask_h))
# Set correct number of bytes per value
mask_array = mask_array.astype(numpy.uint8)
# Convert the mask to bytes
mask_array = mask_array.tostring()
# Pack the bytes to a bit mask
mask_packed = create_mask(mask_array, 1)

# Define mask's fill color
mask_color = ColorHolder()
mask_color.setRed(255)
mask_color.setBlue(0)
mask_color.setGreen(0)
mask_color.setAlpha(100)

# create an ROI with a single mask
mask = omero.model.MaskI()
mask.setTheC(rint(0))
mask.setTheZ(rint(0))
mask.setTheT(rint(0))
mask.setX(rdouble(mask_x))
mask.setY(rdouble(mask_y))
mask.setWidth(rdouble(mask_w))
mask.setHeight(rdouble(mask_h))
mask.setFillColor(rint(mask_color.getInt()))
mask.setTextValue(rstring("test-Mask"))
mask.setBytes(mask_packed)
create_roi(image, [mask])

# create an ROI with single point shape
point = omero.model.PointI()
point.x = rdouble(x)
point.y = rdouble(y)
point.theZ = rint(z)
point.theT = rint(t)
point.textValue = rstring("test-Point")
create_roi(image, [point])


# create an ROI with a single polygon, setting colors and lineWidth
polygon = omero.model.PolygonI()
polygon.theZ = rint(z)
polygon.theT = rint(t)
polygon.fillColor = rint(rgba_to_int(255, 0, 255, 50))
polygon.strokeColor = rint(rgba_to_int(255, 255, 0))
polygon.strokeWidth = omero.model.LengthI(10, UnitsLength.PIXEL)
points = "10,20, 50,150, 200,200, 250,75"
polygon.points = rstring(points)
create_roi(image, [polygon])

# Retrieve ROIs linked to an Image
# ================================
roi_service = conn.getRoiService()
result = roi_service.findByImage(imageId, None)
for roi in result.rois:
    print "ROI:  ID:", roi.getId().getValue()
    for s in roi.copyShapes():
        shape = {}
        shape['id'] = s.getId().getValue()
        shape['theT'] = s.getTheT().getValue()
        shape['theZ'] = s.getTheZ().getValue()
        if s.getTextValue():
            shape['textValue'] = s.getTextValue().getValue()
        if type(s) == omero.model.RectangleI:
            shape['type'] = 'Rectangle'
            shape['x'] = s.getX().getValue()
            shape['y'] = s.getY().getValue()
            shape['width'] = s.getWidth().getValue()
            shape['height'] = s.getHeight().getValue()
        elif type(s) == omero.model.EllipseI:
            shape['type'] = 'Ellipse'
            shape['x'] = s.getX().getValue()
            shape['y'] = s.getY().getValue()
            shape['radiusX'] = s.getRadiusX().getValue()
            shape['radiusY'] = s.getRadiusY().getValue()
        elif type(s) == omero.model.PointI:
            shape['type'] = 'Point'
            shape['x'] = s.getX().getValue()
            shape['y'] = s.getY().getValue()
        elif type(s) == omero.model.LineI:
            shape['type'] = 'Line'
            shape['x1'] = s.getX1().getValue()
            shape['x2'] = s.getX2().getValue()
            shape['y1'] = s.getY1().getValue()
            shape['y2'] = s.getY2().getValue()
        elif type(s) == omero.model.MaskI:
            shape['type'] = 'Mask'
            shape['x'] = s.getX().getValue()
            shape['y'] = s.getY().getValue()
            shape['width'] = s.getWidth().getValue()
            shape['height'] = s.getHeight().getValue()
        elif type(s) in (
                omero.model.LabelI, omero.model.PolygonI):
            print type(s), " Not supported by this code"
        # Do some processing here, or just print:
        print "   Shape:",
        for key, value in shape.items():
            print "  ", key, value,
        print ""


# Remove shape from ROI
# =====================
result = roi_service.findByImage(imageId, None)
for roi in result.rois:
    for s in roi.copyShapes():
        # Find and remove the Shape we added above
        if s.getTextValue() and s.getTextValue().getValue() == "test-Ellipse":
            print "Removing Shape from ROI..."
            roi.removeShape(s)
            roi = updateService.saveAndReturnObject(roi)


# Delete ROIs and all the Shapes they contain
# ===========================================
roi_to_delete = create_roi(image, [rect])
print "Deleting ROI:", roi.getId().getValue()
conn.deleteObjects("Roi", [roi.getId().getValue()], wait=True)


# Close connection
# ================
# When you are done, close the session to free up server resources.
conn.close()

File: examples/Training/matlab/ROIs.m
% Copyright (C) 2011-2014 University of Dundee & Open Microscopy Environment.
% All rights reserved.
%
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License f4or more details.
%
% You should have received a copy of the GNU General Public License along
% with this program; if not, write to the Free Software Foundation, Inc.,
% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
% ROIs

% To learn about the model see  http://www.ome-xml.org/wiki/ROI/2010-04.
% Note that annotation can be linked to ROI.
try
%%
% start-code
%%

    % Initialize a client and a session using the ice.config file
    % See ConnectToOMERO for alternative ways to initialize a session
    [client, session] = loadOmero();
    p = parseOmeroProperties(client);
    eventContext = session.getAdminService().getEventContext();
    fprintf(1, 'Created connection to %s\n', p.hostname);
    msg = 'Created session for user %s (id: %g) using group %s (id: %g)\n';
    fprintf(1, msg, char(eventContext.userName), eventContext.userId,...
        char(eventContext.groupName), eventContext.groupId);
  
    % Information to edit
    imageId = p.imageid;
    
    % Load image
    image = getImages(session, imageId);
    assert(~isempty(image), 'OMERO:ROIs', 'Image id not valid');
    
% Create shapes
% =============

    % Create a rectangle
    disp('Create rectangular shape');
    rectangle = createRectangle(0, 0, 10, 20);
    rectangle.setFillColor(rint(hex2dec('7DFF0000')));    % 'aRGB' - 50% red
    rectangle.setStrokeColor(rint(hex2dec('FF000000')));  % 'aRGB' - 100% black
    rectangle.setTextValue(rstring('Rectangle'));
    % indicate on which plane to attach the shape
    rectangle = setShapeCoordinates(rectangle, 0, 0, 0);

    % Create an ellipse
    disp('Create ellipsoidal shape');
    ellipse = createEllipse(10, 10, 10, 10);
    ellipse.setFillColor(rint(hex2dec('7D0000FF')));    % 'aRGB' - 50% blue
    ellipse.setStrokeColor(rint(hex2dec('FF000000')));  % 'aRGB' - 100% black
    ellipse.setTextValue(rstring('Ellipse'));
    % indicate on which plane to attach the shape
    ellipse = setShapeCoordinates(ellipse, 0, 0, 1);
    
    % Create a point
    disp('Create point shape');
    point = createPoint(5, 4);
    point.setTextValue(rstring('Point'));
    % indicate on which plane to attach the shape
    point = setShapeCoordinates(point, 0, 0, 2);
    
    % Create a line
    disp('Create line shape');
    line = createLine([10 15], [10 20]);
    line.setStrokeColor(rint(hex2dec('FF000000')));  % 'aRGB' - 100% black
    line.setTextValue(rstring('Line'));
    % indicate on which plane to attach the shape
    line = setShapeCoordinates(line, 0, 0, 3);
    
    % Create a polyline
    disp('Create polyline shape');
    polyline = createPolyline([1 5 10 8], [1 5 5 10]);
    polyline.setStrokeColor(rint(hex2dec('FF000000')));  % 'aRGB' - 100% black
    polyline.setTextValue(rstring('Polyline'));
    % indicate on which plane to attach the shape
    polyline = setShapeCoordinates(polyline, 0, 0, 4);
    
    % Create a polygon
    disp('Create polygon shape');
    polygon = createPolygon([1 5 10 8], [1 5 5 10]);
    polygon.setStrokeColor(rint(hex2dec('FF000000')));  % 'aRGB' - 100% black
    polygon.setTextValue(rstring('Polygon'));
    % indicate on which plane to attach the shape
    polygon = setShapeCoordinates(polygon, 0, 0, 5);
    
    % Create the roi.
    roi = omero.model.RoiI;
    % Attach the shape to the roi, several shapes can be added.
    roi.addShape(rectangle);
    roi.addShape(ellipse);
    roi.addShape(point);
    roi.addShape(line);
    roi.addShape(polyline);
    roi.addShape(polygon);
    % Link the roi and the image
    roi.setImage(omero.model.ImageI(imageId, false));
    % Save the ROI
    roi = session.getUpdateService().saveAndReturnObject(roi);
    fprintf(1, 'Created ROI %g\n', roi.getId().getValue());
    
    % Check that the shapes have been added.
    fprintf(1, 'Reading shapes attached to ROI %g\n', roi.getId().getValue());
    nShapes = roi.sizeOfShapes;
    fprintf(1, 'Found %g shapes\n', nShapes);
    for i = 1 : nShapes
        shape = roi.getShape(i - 1);
        if (isa(shape, 'omero.model.Rectangle'))
            %handle rectangle
            x = shape.getX().getValue();
            y = shape.getY().getValue();
            width = shape.getWidth().getValue();
            height = shape.getHeight().getValue();
            fprintf(1, '  Rectangle X: %g, Y: %g, Width: %g, Height: %g\n',...
                x, y, width, height);      
        elseif (isa(shape, 'omero.model.Ellipse'))
            x = shape.getX().getValue();
            y = shape.getY().getValue();
            radiusx = shape.getRadiusX().getValue();
            radiusy = shape.getRadiusY().getValue();
            fprintf(1, '  Ellipse X: %g, Y: %g, RadiusX: %g, RadiusY: %g\n',...
                x, y, radiusx, radiusy);
        elseif (isa(shape, 'omero.model.Point'))
            x = shape.getX().getValue();
            y = shape.getY().getValue();
            fprintf(1, '  Point X: %g, Y: %g\n', x, y);
        elseif (isa(shape, 'omero.model.Line'))
            x1 = shape.getX1().getValue();
            x2 = shape.getX2().getValue();
            y1 = shape.getY1().getValue();
            y2 = shape.getY2().getValue();
            fprintf(1, '  Line (X1, Y1): (%g, %g), (X2, Y2): (%g, %g)\n',...
                x1, y1, x2, y2);
        elseif isa(shape, 'omero.model.Polyline')
            points = shape.getPoints().getValue();
            fprintf(1, '  Poyline: %s\n', char(points));
        elseif (isa(shape, 'omero.model.Polygon'))
            points = shape.getPoints().getValue();
            fprintf(1, '  Polygon: %s\n', char(points));
        end
    end

    % Create a mask covering half of the image
    pixels = image.getPrimaryPixels();
    sizeX = pixels.getSizeX().getValue();
    sizeY = pixels.getSizeY().getValue();
    m = false(sizeY, sizeX / 2);
    m(end/4:3*end/4,end/2:end)=true;
    mask = createMask(m);
    setShapeCoordinates(mask, 0, 0, 0);
    roi = omero.model.RoiI();
    roi.addShape(mask);
    roi.setImage(omero.model.ImageI(imageId, false));
    roi = session.getUpdateService().saveAndReturnObject(roi);
    fprintf(1, 'Created ROI %g\n', roi.getId().getValue());
    
    % Create a mask shape at a position different from (0, 0)
    m = true(sizeY/8 - 4, sizeX/8 - 4);
    m(end/4 : 3 * end/4, end/4 : 3 * end/4) = false;
    mask = createMask(5 * sizeY / 8, 5 * sizeX/8, m);
    setShapeCoordinates(mask, 0, 0, 0);
    roi = omero.model.RoiI();
    roi.addShape(mask);
    roi.setImage(omero.model.ImageI(imageId, false));
    roi = session.getUpdateService().saveAndReturnObject(roi);
    fprintf(1, 'Created ROI %g\n', roi.getId().getValue());

    % Create and apply transforms to shapes and add them to ROI
    
    % Apply shearing alone to a rectangle (Shear factor set to 10 in the x
    % and y dimensions in this example)
    % create rectangle (shape object)
    rectangle = createRectangle(0, 0, 10, 20);
    rectangle = setShapeCoordinates(rectangle, 0, 0, 0);
    % x and y shear factors
    xShear = 10;
    yShear = 10;
    % create transform object
    newTform = omero.model.AffineTransformI;
    newTform.setA00(rdouble(1));
    newTform.setA10(rdouble(yShear));
    newTform.setA01(rdouble(xShear));
    newTform.setA11(rdouble(1));
    newTform.setA02(rdouble(0));
    newTform.setA12(rdouble(0));
    % apply transform
    rectangle.setTransform(newTform);
    
    % Apply rotation alone to an ellipse object
    % (angle of rotation set to 10 degrees)
    % create ellipse (shape object)
    ellipse = createEllipse(0, 0, 10, 20);
    setShapeCoordinates(ellipse, 0, 0, 0);
    % set angle of rotation
    theta = 10;
    % create transform object
    newTform = omero.model.AffineTransformI;
    newTform.setA00(rdouble(cos(theta)));
    newTform.setA10(rdouble(-sin(theta)));
    newTform.setA01(rdouble(sin(theta)));
    newTform.setA11(rdouble(cos(theta)));
    newTform.setA02(rdouble(0));
    newTform.setA12(rdouble(0));
    % apply transform
    ellipse.setTransform(newTform);
    
    % Apply translation alone to a point object (10 pixels in the x and y
    % direction)
    % create point (shape object)
    point = createPoint(5, 4);
    setShapeCoordinates(point, 0, 0, 0);
    % translation parameters
    xTranslate = 10;
    yTranslate = 10;
    % create transform object
    newTform = omero.model.AffineTransformI;
    newTform.setA00(rdouble(1));
    newTform.setA10(rdouble(0));
    newTform.setA01(rdouble(0));
    newTform.setA11(rdouble(1));
    newTform.setA02(rdouble(xTranslate));
    newTform.setA12(rdouble(yTranslate));
    % apply transform
    point.setTransform(newTform);
    
    % Apply a scale change alone to a polygon object (set scale factor to 10
    % in the x  and y direction)
    % create polygon (shape object)
    polygon = createPolygon([1 5 10 8], [1 5 5 10]);
    setShapeCoordinates(polygon, 0, 0, 0);
    % scaling parameters
    xScale = 10;
    yScale = 10;
    % create transform object
    newTform = omero.model.AffineTransformI;
    newTform.setA00(rdouble(xScale));
    newTform.setA10(rdouble(0));
    newTform.setA01(rdouble(0));
    newTform.setA11(rdouble(yScale));
    newTform.setA02(rdouble(0));
    newTform.setA12(rdouble(0));
    % apply transform
    polygon.setTransform(newTform);
    
    % Create the roi.
    roi = omero.model.RoiI;
    % Attach the shape to the roi, several shapes can be added.
    roi.addShape(rectangle);
    roi.addShape(ellipse);
    roi.addShape(point);
    roi.addShape(polygon);
    % Link the roi and the image
    roi.setImage(omero.model.ImageI(imageId, false));
    % Save the ROI
    roi = session.getUpdateService().saveAndReturnObject(roi);
    fprintf(1, 'Created ROIs with Transform objects %g\n', roi.getId().getValue());
    
    % Check that the transforms have been added to the shapes.
    fprintf(1, 'Reading shapes attached to ROI %g\n', roi.getId().getValue());
    nShapes = roi.sizeOfShapes;
    fprintf(1, 'Found %g shapes\n', nShapes);
    for i = 1 : nShapes
        shape = roi.getShape(i - 1);
        
        %http://blog.openmicroscopy.org/data-model/future-plans/2016/06/20/shape-transforms/
        transform = shape.getTransform;
        xScaling = transform.getA00.getValue;
        xShearing = transform.getA01.getValue;
        xTranslation = transform.getA02.getValue;
            
        yScaling = transform.getA11.getValue;
        yShearing = transform.getA10.getValue;
        yTranslation = transform.getA12.getValue;
        
        %tformMatrix = [A00, A10, 0; A01, A11, 0; A02, A12, 1];
        tformMatrix = [xScaling, yShearing, 0; xShearing, yScaling, 0; xTranslation, yTranslation, 1];
        
        fprintf(1, 'Shape Type : %s\n', char(shape.toString));
        fprintf(1, 'xScaling : %s\n', num2str(tformMatrix(1,1)));
        fprintf(1, 'yScaling : %s\n', num2str(tformMatrix(2,2)));
        fprintf(1, 'xShearing : %s\n', num2str(tformMatrix(2,1)));
        fprintf(1, 'yShearing : %s\n', num2str(tformMatrix(1,2)));
        fprintf(1, 'xTranslation: %s\n', num2str(tformMatrix(3,1)));
        fprintf(1, 'yTranslation: %s\n', num2str(tformMatrix(3,2)));
    end
%%
% end-code
%%

catch err
    client.closeSession();
    throw(err);
end

% Close the session
client.closeSession();
